bee-network/bee-autopeering/src/server.rs

```diff
@@ -0,0 +1,290 @@
+                        }
+                        Err(e) => {
+                            log::error!("UDP socket read error; stopping incoming packet handler. Cause: {}", e);
+                            // TODO: intiate shutdown of the system
+                            break 'recv;
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Runnable for OutgoingPacketHandler {
+    const NAME: &'static str = "OutgoingPacketHandler";
+    const SHUTDOWN_PRIORITY: u8 = 3;
+
+    type ShutdownSignal = ShutdownRx;
+
+    async fn run(self, mut shutdown_rx: Self::ShutdownSignal) {
+        let OutgoingPacketHandler {
+            outgoing_socket,
+            mut outgoing_rx,
+            local,
+            bind_addr,
+        } = self;
+
+        'recv: loop {
+            tokio::select! {
+                _ = &mut shutdown_rx => {
+                    break;
+                }
+                o = outgoing_rx.recv() => {
+                    if let Some(packet) = o {
+                        let OutgoingPacket {
+                            msg_type,
+                            msg_bytes,
+                            peer_addr,
+                        } = packet;
+
+                        if peer_addr == bind_addr {
+                            log::warn!("Trying to send to own bind address: {}. Ignoring packet.", peer_addr);
+                            continue 'recv;
+                        }
+
+                        let marshalled_bytes = marshal(msg_type, &msg_bytes);
+
+                        let signature = local.sign(&marshalled_bytes);
+                        let packet = Packet::new(msg_type, &marshalled_bytes, &local.public_key(), signature);
+
+                        let bytes = packet.to_protobuf().expect("error encoding outgoing packet");
+
+                        if bytes.len() > MAX_PACKET_SIZE {
+                            log::warn!("Trying to send too many bytes to {}. Ignoring...", peer_addr);
+                            continue 'recv;
+                        }
+
+                        let n = outgoing_socket.send_to(&bytes, peer_addr).await.expect("socket send error");
+
+                        log::trace!("Sent {} bytes to {}.", n, peer_addr);
+                    } else {
+                        // All `outgoing_tx` message senders were dropped.
+                        break 'recv;
+                    }
+                }
+            }
+        }
+    }
+}
+
+pub(crate) fn marshal(msg_type: MessageType, msg_bytes: &[u8]) -> Vec<u8> {
+    let mut marshalled_bytes = vec![0u8; msg_bytes.len() + 1];
+    marshalled_bytes[0] = msg_type as u8;
+    marshalled_bytes[1..].copy_from_slice(msg_bytes);
+    marshalled_bytes
+}
+
+pub(crate) fn unmarshal(marshalled_bytes: &[u8]) -> (MessageType, Vec<u8>) {
+    let msg_type = num::FromPrimitive::from_u32(marshalled_bytes[0] as u32).expect("unknown message type");
```

Comment:
same idea about splitting

Reply:
Same question as above. I don't see how things become more efficient, but I might just be able anymore to follow.

Reply:
Closing this for now (added a comment as a reminder for Christian in case he wants to make a PR to optimize this they way he means to)